cloud/amazon/ec2_vpc_dhcp_options.py

```diff
@@ -128,105 +159,207 @@
-    # for data validation, but still need to cast it to a string
-    new_options['netbios-node-type'] = _get_updated_option(
-            [str(params['netbios_node_type'])], existing_options.options.get('netbios-node-type'),
-            inherit_existing)
-  
-    new_options['domain-name'] = _get_updated_option(
-            [params['domain_name']], existing_options.options.get('domain-name'), inherit_existing)
-  
-    if existing_options and new_options == existing_options.options:
-        module.exit_json(changed=False)
-  
-    if new_options['netbios-node-type']:
-        new_options['netbios-node-type'] = new_options['netbios-node-type'][0]
-  
-    if new_options['domain-name']:
-        new_options['domain-name'] = new_options['domain-name'][0]
-  
-    if not module.check_mode:
-        dhcp_option = connection.create_dhcp_options(new_options['domain-name'],
-                new_options['domain-name-servers'], new_options['ntp-servers'],
-                new_options['netbios-name-servers'], new_options['netbios-node-type'])
+
+    # First check if we were given a dhcp_options_id
+    if not params['dhcp_options_id']:
+    # No, so create new_options from the parameters
+        if params['dns_servers'] != None:
+            new_options['domain-name-servers'] = params['dns_servers']
+        if params['netbios_name_servers'] != None:
+            new_options['netbios-name-servers'] = params['netbios_name_servers']
+        if params['ntp_servers'] != None:
+            new_options['ntp-servers'] = params['ntp_servers']
+        if params['domain_name'] != None:
+            # needs to be a list for comparison with boto objects later
+            new_options['domain-name'] = [ params['domain_name'] ]
+        if params['netbios_node_type'] != None:
+            # needs to be a list for comparison with boto objects later
+            new_options['netbios-node-type'] = [ str(params['netbios_node_type']) ]
+        # If we were given a vpc_id then we need to look at the options on that
+        if params['vpc_id']:
+            existing_options = fetch_dhcp_options_for_vpc(connection, params['vpc_id'])
+            # if we've been asked to inherit existing options, do that now
+            if params['inherit_existing']:
+                if not existing_options:
+                    module.fail_json(msg="a vpc_id was provided, but that vpc does not exist")
+                for option in [ 'domain-name-servers', 'netbios-name-servers', 'ntp-servers', 'domain-name', 'netbios-node-type']:
+                    if existing_options.options.get(option) and (not new_options[option] or [''] == new_options[option]):
+                        new_options[option] = existing_options.options.get(option)
+    
+        # Do the vpc's dhcp options already match what we're asked for? if so we are done
+        if existing_options and new_options == existing_options.options:
```

Comment:
I think this has the potential to raise an exception -- `existing_options` is only defined if `vpc_id` is passed in as a parameter. If nether dhcp_options_id nor vpc_id is passed in, then this will error out.